### ISSUE TITLE
Metrics with naming issues

### ISSUE THREAD

**Issue by SuperQ** [prometheus/prometheus#8718]
* Prometheus version:

2.26.0

* Problem

There appear to be some metrics that don't follow the [metric naming best practices](https://prometheus.io/docs/practices/naming/#metric-names).

Most of these appear to be related to `_failed_total` vs `failed_unit_total`. The best practices doc, and the consensus has been that the unit should be the last word in the metric name (besides total/count/sum).

1. Update the naming best practices to include a "unit modifier" like "failed".
2. Update the metric names like `prometheus_tsdb_head_truncations_failed_total` to be `prometheus_tsdb_head_failed_truncations_total`.

**Comment by roidelapluie** [comment]
This needs more discussion.

I think that  prometheus_tsdb_head_truncations_failed_total is better than prometheus_tsdb_head_failed_truncations_total.

I just see a gap in the docs to document the as-is situation.

**Comment by roidelapluie** [comment]
It also occurs to me that it is easier to compare:

net_conntrack_dialer_conn_failed_total
net_conntrack_dialer_conn_established_total
net_conntrack_dialer_conn_closed_total
net_conntrack_dialer_conn_attempted_total

Than

net_conntrack_dialer_failed_conn_total
net_conntrack_dialer_established_conn_total
net_conntrack_dialer_closed_conn_total
net_conntrack_dialer_attempted_conn_total

**Comment by roidelapluie** [comment]
Same with

prometheus_target_scrape_pools_failed_total
prometheus_target_scrape_pools_total

and

prometheus_target_scrape_failed_pools_total
prometheus_target_scrape_pools_total


It is easier to work with the first ones.

**Comment by grobie** [comment]
Thanks for starting this discussion and providing these examples @roidelapluie. This Github issue is the result of a similar discussion @SuperQ and I had in https://github.com/prometheus/memcached_exporter/pull/109. I was surprised to learn that there was apparently a consensus for a new metric naming pattern as Ben states, I hadn't heard of such discussion before. I have always chosen metric names where similar metrics are grouped together, instead of moving a modifier to the beginning and thus making it harder to list metrics as shown in @roidelapluie's examples.

@SuperQ Do you remember where and with whom you had these discussions, so that we can read up on the previous arguments?

**Comment by SuperQ** [comment]
I don't remember. Basically, there's a lack of consistency in how to deal with these cases. My understanding was we settled on unit as the most-right word within the metric. But is `foo_failed` a composite unit? I don't know.

**Comment by juliusv** [comment]
As the one who wrote the naming conventions page, I'm for keeping the metrics grouped as they are in this case, because I don't see e.g. `truncations` as a unit, but more as a dimension (not in the sense that it needs to be a label, but in the sense that it describes what is being measured, and not which units it's being measured in). The grouping might be awkward to read initially, but it keeps metrics that belong together in the same underscored hierarchy, easy to find and work with.

**Comment by roidelapluie** [comment]
It seems that the general consensus is that the current metrics names are correct. Should we close this issue @SuperQ ?

**Comment by SuperQ** [comment]
I'm still not sure how "dimension" is not a unit. 

https://www.dictionary.com/browse/unit

> 6. Also called dimension. any specified amount of a quantity, as of length, volume, force, momentum, or time, by comparison with which any other quantity of the same kind is measured or estimated.

By this definition, `truncations` is a unit.

For the subject of "composite units", I'm fine with `truncations_failed` over `failed_turncations`. It would be good to mention this case in the docs.

**Comment by juliusv** [comment]
When I talk about "dimension", I mean the first definition listed in https://www.dictionary.com/browse/dimension. A dimension in that sense is a degree of freedom (length, time, weight, ...) that can be measured separately from other dimensions, but it doesn't say anything about the unit of measurement yet. A unit of measurement can then be used to specify the precise magnitude of a value along a dimension.

With `truncations` and other countable quantities, the word itself can arguably be seen as both a dimension and a unit (or you could even say that it's unitless), but I would say that it's first and foremost a dimension (because it specifies the aspect of a system that is being measured).

So yeah, I prefer `truncations_failed` due to the lexicographic sorting + discoverability aspect, but `failed_truncations` wouldn't be wrong either.

**Comment by grobie** [comment]
I agree with @juliusv, I would not consider every countable thing in the world to be a unit. Regardless I also strongly believe in the advantages of lexicographic sorting.

The current documentation does not provide much guidance or examples in naming. So instead of just closing this issue, let's improve it and add more examples from our discussion here?
